src/aiokatcp/sensor.py

```diff
@@ -226,11 +236,19 @@ def params(self) -> List[bytes]:
         else:
             return []
 
-    def attach(self, observer: Callable[["Sensor[_T]", Reading[_T]], None]) -> None:
-        self._observers.add(observer)
+    def attach(self, observer: Observer) -> None:
+        sig = inspect.signature(observer)
+        if "old_reading" in sig.parameters:
+            self._change_aware_observers.add(observer)  # type: ignore
+        else:
+            self._classic_observers.add(observer)  # type: ignore
 
-    def detach(self, observer: Callable[["Sensor[_T]", Reading[_T]], None]) -> None:
-        self._observers.discard(observer)
+    def detach(self, observer: Observer) -> None:
+        sig = inspect.signature(observer)
+        if "old_reading" in sig.parameters:
+            self._change_aware_observers.discard(observer)  # type: ignore
+        else:
+            self._classic_observers.discard(observer)  # type: ignore
```

Comment:
It'll be simpler (and quite likely faster, given how intricate I expect `inspect.signature` to be) to just discard it from both sets (`discard` is a no-op if no present).